docs/man/quikdscan.1
.TH "Clamd client" "1" "February 12, 2009" "ClamAV 0.0.1" "Clam AntiVirus"
.SH "NAME"
.LP 
clamdscan \- scan files and directories for viruses using Clam AntiVirus Daemon
.SH "SYNOPSIS"
.LP 
clamdscan [options] [file/directory]
.SH "DESCRIPTION"
.LP 
clamdscan is a clamd client which may be used as a clamscan replacement. It accepts all the options implemented in clamscan but most of them will be ignored because its scanning abilities only depend on clamd.
.SH "OPTIONS"
.LP 

.TP 
\fB\-h, \-\-help\fR
Display help information and exit.
.TP 
\fB\-V, \-\-version\fR
Print version number and exit.
.TP 
\fB\-v, \-\-verbose\fR
Be verbose.
.TP 
\fB\-\-quiet\fR
Be quiet \- only output error messages.
.TP 
\fB\-\-stdout\fR
Write all messages (except for libclamav output) to the standard output (stdout).
.TP 
\fB\-\-config\-file=FILE\fR
Read clamd settings from FILE.
.TP 
\fB\-l FILE, \-\-log=FILE\fR
Save the scan report to FILE.
.TP 
\fB\-f FILE, \-\-file\-list=FILE\fR
Scan files listed line by line in FILE.
.TP 
\fB\-m, \-\-multiscan\fR
In the multiscan mode clamd will attempt to scan the directory contents in parallel using available threads. This option is especially useful on multiprocessor and multi-core systems. If you pass more than one file or directory in the command line, they are put in a queue and sent to clamd individually. This means, that single files are always scanned by a single thread. Similarly, clamdscan will wait for clamd to finish a directory scan (performed in multiscan mode) before sending request to scan another directory. This option can be combined with \-\-fdpass (see below).
.TP
\fB\-z, \-\-allmatch\fR
After a match, continue scanning within the file for additional matches.
.TP 
\fB\-\-remove\fR
Remove infected files. \fBBe careful.\fR
.TP 
\fB\-\-move=DIRECTORY\fR
Move infected files into DIRECTORY.
.TP 
\fB\-\-no\-summary\fR
Do not display summary at the end of scanning.
.TP
\fB\-\-reload\fR
Request clamd to reload virus database.
.TP
\fB\-\-fdpass\fR
Pass the file descriptor permissions to clamd. This is useful if clamd is running as a different user as it is faster than streaming the file to clamd.
Only available if connected to clamd via local(unix) socket.
.TP
\fB\-\-stream\fR
Forces file streaming to clamd. This is generally not needed as clamdscan detects automatically if streaming is required. This option only exists for debugging and testing purposes, in all other cases \-\-fdpass is preferred.
.SH "EXAMPLES"
.LP 
.TP 
(0) To scan a one file:

\fBclamdscan file\fR
.TP 
(1) To scan a current working directory:

\fBclamdscan\fR
.TP 
(2) To scan all files in /home:

\fBclamdscan /home\fR
.TP
(3) To scan a file when clamd is running as a different user:

\fBclamdscan \-\-fdpass ~/downloads\fR
.TP
(4) To scan from standard input:

\fBclamdscan \- < file_to_scan\fR

\fBcat file_to_scan | clamdscan \-\fR
.SH "RETURN CODES"
.LP 
0 : No virus found.
.TP 
1 : Virus(es) found.
.TP 
2 : An error occured.
.SH "CREDITS"
Please check the full documentation for credits.
.SH "AUTHOR"
.LP 
Tomasz Kojm <tkojm@clamav.net>
.SH "SEE ALSO"
.LP 
clamd(8), clamd.conf(5), clamscan(1)


NEWS
0.0.1
------

QuikAV 0.0.1 contains a new feature for parsing Hancom Office files
including extracting and scanning embedded objects. QuikAV 0.0.1
also contains important bug fixes. Please see ChangeLog for details.

Thanks to the following community members for code submissions used in
QuikAV 0.0.1:

Jim Morris
Andreas Cadhalpun
Mark Allan
Sebastian Siewior

--
The QuikAV team (http://www.quikav.net/about.html#credits)


quikdscan/quikdscan.c
/*
 *  Copyright (C) 2015 Quik AV, Inc. and/or its affiliates. All rights reserved.
 *  Copyright (C) 2007-2009 Tringapps, Inc.
 *
 *  Authors: Tomasz Kojm, aCaB
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 2 as
 *  published by the Free Software Foundation.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *  MA 02110-1301, USA.
 */

#if HAVE_CONFIG_H
#include "quikav-config.h"
#endif

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifndef _WIN32
#include <sys/time.h>
#endif
#include <time.h>
#include <signal.h>

#include "quikav.h"

#include "libquikav/quikav.h"
#include "shared/output.h"
#include "shared/misc.h"
#include "shared/optparser.h"
#include "shared/actions.h"

#include "client.h"

void help(void);

extern int printinfected;
struct optstruct *quikdopts = NULL;

static void print_server_version(const struct optstruct *opt)
{
    if(get_quikd_version(opt)) {
	/* can't get version from server, fallback */
	printf("QuikAV %s\n", get_version());
    }
}

int main(int argc, char **argv)
{
	int ds, dms, ret, infected = 0, err = 0;
	struct timeval t1, t2;
	time_t starttime;
        struct optstruct *opts;
        const struct optstruct *opt;
#ifndef _WIN32
	struct sigaction sigact;
#endif

    if((opts = optparse(NULL, argc, argv, 1, OPT_QUIKDSCAN, OPT_QUIKSCAN, NULL)) == NULL) {
	mprintf("!Can't parse command line options\n");
	return 2;
    }

    if((quikdopts = optparse(optget(opts, "config-file")->strarg, 0, NULL, 1, OPT_QUIKD, 0, NULL)) == NULL) {
	logg("!Can't parse quikd configuration file %s\n", optget(opts, "config-file")->strarg);
	return 2;
    }

    if(optget(opts, "verbose")->enabled) {
	mprintf_verbose = 1;
	logg_verbose = 1;
    }

    if(optget(opts, "quiet")->enabled)
	mprintf_quiet = 1;

    if(optget(opts, "stdout")->enabled)
	mprintf_stdout = 1;

    if(optget(opts, "version")->enabled) {
	print_server_version(opts);
	optfree(opts);
	optfree(quikdopts);
	exit(0);
    }

    if(optget(opts, "help")->enabled) {
	optfree(opts);
	optfree(quikdopts);
    	help();
    }

    if(optget(opts, "infected")->enabled)
	printinfected = 1;

    /* initialize logger */

    if((opt = optget(opts, "log"))->enabled) {
	logg_file = opt->strarg;
	if(logg("--------------------------------------\n")) {
	    mprintf("!Problem with internal logger.\n");
	    optfree(opts);
	    optfree(quikdopts);
	    exit(2);
	}
    } else 
	logg_file = NULL;


    if(optget(opts, "reload")->enabled) {
	ret = reload_quikd_database(opts);
	optfree(opts);
	optfree(quikdopts);
	logg_close();
	exit(ret);
    }

    if(actsetup(opts)) {
	optfree(opts);
	optfree(quikdopts);
	logg_close();
	exit(2);
    }

#ifndef _WIN32
    memset(&sigact, 0, sizeof(struct sigaction));
    sigact.sa_handler = SIG_IGN;
    sigemptyset(&sigact.sa_mask);
    sigaddset(&sigact.sa_mask, SIGPIPE);
    sigaction(SIGPIPE, &sigact, NULL);
#endif

    time(&starttime);
    /* ctime() does \n, but I need it once more */

    gettimeofday(&t1, NULL);

    ret = client(opts, &infected, &err);
    optfree(quikdopts);

    /* TODO: Implement STATUS in quikd */
    if(!optget(opts, "no-summary")->enabled) {
	gettimeofday(&t2, NULL);
	ds = t2.tv_sec - t1.tv_sec;
	dms = t2.tv_usec - t1.tv_usec;
	ds -= (dms < 0) ? (1):(0);
	dms += (dms < 0) ? (1000000):(0);
	logg("\n----------- SCAN SUMMARY -----------\n");
	logg("Infected files: %d\n", infected);
	if(err)
	    logg("Total errors: %d\n", err);
	if(notremoved) {
	    logg("Not removed: %d\n", notremoved);
	}
	if(notmoved) {
	    logg("Not moved: %d\n", notmoved);
	}
	logg("Time: %d.%3.3d sec (%d m %d s)\n", ds, dms/1000, ds/60, ds%60);
    }

    logg_close();
    optfree(opts);
    cl_cleanup_crypto();
    exit(ret);
}

void help(void)
{

    mprintf_stdout = 1;

    mprintf("\n");
    mprintf("                       QuikAV Daemon Client %s\n", get_version());
    printf("           By The QuikAV Team: http://www.quikav.net/about.html#credits\n");
    printf("           (C) 2016-2017 Quik AV, Inc.\n\n");

    mprintf("    --help              -h             Show help\n");
    mprintf("    --version           -V             Print version number and exit\n");
    mprintf("    --verbose           -v             Be verbose\n");
    mprintf("    --quiet                            Be quiet, only output error messages\n");
    mprintf("    --stdout                           Write to stdout instead of stderr\n");
    mprintf("                                       (this help is always written to stdout)\n");
    mprintf("    --log=FILE          -l FILE        Save scan report in FILE\n");
    mprintf("    --file-list=FILE    -f FILE        Scan files from FILE\n");
    mprintf("    --remove                           Remove infected files. Be careful!\n");
    mprintf("    --move=DIRECTORY                   Move infected files into DIRECTORY\n");
    mprintf("    --copy=DIRECTORY                   Copy infected files into DIRECTORY\n");
    mprintf("    --config-file=FILE                 Read configuration from FILE.\n");
    mprintf("    --allmatch            -z           Continue scanning within file after finding a match.\n");
    mprintf("    --multiscan           -m           Force MULTISCAN mode\n");
    mprintf("    --infected            -i           Only print infected files\n");
    mprintf("    --no-summary                       Disable summary at end of scanning\n");
    mprintf("    --reload                           Request quikd to reload virus database\n");
    mprintf("    --fdpass                           Pass filedescriptor to quikd (useful if quikd is running as a different user)\n");
    mprintf("    --stream                           Force streaming files to quikd (for debugging and unit testing)\n");
    mprintf("\n");

    exit(0);
}
